### PR TITLE
CUDA backward bug

### PR DESCRIPTION
**Pull request by frostedoyster**
At the moment, we decide whether to calculate double backward second derivatives by checking if `ddsph` is defined or not. This is fine for the CPU code, but on the CUDA code `dsph` and `ddsph` are defined also if these are not required. This leads to second derivatives of the wrong shape which cause torch to crash later on

### CODE REVIEW

**Comment by frostedoyster** [comment]
I don't know how to ping this, but the problematic parts are in the CUDA file, lines 723-751:
`    torch::Tensor d_sph;
    if (xyz.requires_grad() || gradients)
    {
        d_sph = torch::empty(
            {xyz.size(0), 3, n_total},
            torch::TensorOptions().dtype(xyz.dtype()).device(xyz.device()));
    }
    else
    {
        // just so accessor doesn't complain
        d_sph = torch::empty(
            {1, 1, 1},
            torch::TensorOptions().dtype(xyz.dtype()).device(xyz.device()));
    }

    torch::Tensor hess_sph;

    if (xyz.requires_grad() && hessian)
    {
        hess_sph = torch::empty(
            {xyz.size(0), 3, 3, n_total},
            torch::TensorOptions().dtype(xyz.dtype()).device(xyz.device()));
    }
    else
    {
        hess_sph = torch::empty(
            {1, 1, 1, 1},
            torch::TensorOptions().dtype(xyz.dtype()).device(xyz.device()));
    }`

**Review by Luthaf — APPROVED** [review]
The fix looks fine
